File: jsol-compiler-py/1020-js-params.py
import math
import functools
from jsol_core import JSOL

# @JSOL v0.2.97
def fProcessParams(saCode): 

  saResult = saCode;
  bContinue = True;
  iOffset = 0;

  while bContinue == True: 

    iSearchLen = len(saResult) - iOffset;
    if iSearchLen <= 0: 

      bContinue = False; continue;


    saSearchArea = saResult[( iOffset):( iOffset)+( iSearchLen)];
    iRelIdx = JSOL.str_index_of(saSearchArea,  "function");
    if iRelIdx == -1: 

      bContinue = False; continue;


    iStartIdx = iOffset + iRelIdx;
    iParenScan = iStartIdx + 8;
    iRLen = len(saResult);

    while iParenScan < iRLen and (saResult[( iParenScan):( iParenScan)+( 1)] == " " or saResult[( iParenScan):( iParenScan)+( 1)] == "\t" or saResult[( iParenScan):( iParenScan)+( 1)] == "\n" or saResult[( iParenScan):( iParenScan)+( 1)] == "\r"): 

      iParenScan = iParenScan + 1;


    if iParenScan >= iRLen or saResult[( iParenScan):( iParenScan)+( 1)] != "(": 

      iOffset = iStartIdx + 8;
      continue;


    iCloseParen = iComp_FindCloseParen(saResult, iParenScan);
    if iCloseParen == -1: 

      bContinue = False; continue;


    saRawParams = saResult[( iParenScan + 1):( iParenScan + 1)+( iCloseParen - iParenScan - 1)];
    saTrimmedParams = saRawParams.strip();
    saTypedParams = "";

    if len(saTrimmedParams) > 0: 

      aParts = JSOL.str_split(saTrimmedParams,  ",");
      iPartsCount = len(aParts);
      aTypedParts = [];
      iP = 0;
      while iP < iPartsCount: 

        saRawPart = aParts[iP].strip();
        saTypedPart = saRawPart;
        if len(saRawPart) > 0 and JSOL.str_index_of(saRawPart,  ":") == -1: 

          saTypedPart = saRawPart + ": any";


        aTypedParts.append( saTypedPart);

        iP = iP + 1;


      saTypedParams =  ", ".join(str(_x) for _x in aTypedParts);


    saBefore = saResult[( 0):( 0)+( iParenScan + 1)];
    saAfter = saResult[( iCloseParen):( iCloseParen)+( len(saResult) - iCloseParen)];

    saResult = saBefore + "" + saTypedParams + "" + saAfter;
    iOffset = iParenScan + 1 + len(saTypedParams) + 1;


  return saResult;

File: jsol-compiler-ts/1020-js-params.ts
declare var JSOL: any;
declare var Rgx: any;
declare var Str: any;
declare var Arr: any;
declare var Bool: any;
declare var Cast: any;

// @JSOL v0.2.97
const $fProcessParams = function ($saCode: any) {
  let $saResult: string = $saCode;
	let $bContinue: boolean = true;
	let $iOffset: number = 0;

	while ($bContinue === true) {
    const $iSearchLen: number = Str["len"]($saResult) - $iOffset;
		if ($iSearchLen <= 0) {
      $bContinue = false; continue;
    }
    const $saSearchArea: string = Str["sub"]($saResult,  $iOffset,  $iSearchLen);
		const $iRelIdx: number = Str["indexOf"]($saSearchArea,  "function");
		if ($iRelIdx === -1) {
      $bContinue = false; continue;
    }
    const $iStartIdx: number = $iOffset + $iRelIdx;
		let $iParenScan: number = $iStartIdx + 8;
		const $iRLen: number = Str["len"]($saResult);

		while ($iParenScan < $iRLen && (Str["sub"]($saResult,  $iParenScan,  1) === " " || Str["sub"]($saResult,  $iParenScan,  1) === "\t" || Str["sub"]($saResult,  $iParenScan,  1) === "\n" || Str["sub"]($saResult,  $iParenScan,  1) === "\r")) {
      $iParenScan = $iParenScan + 1;
    }
    if ($iParenScan >= $iRLen || Str["sub"]($saResult,  $iParenScan,  1) !== "(") {
      $iOffset = $iStartIdx + 8;
			continue;
    }
    const $iCloseParen: number = $iComp_FindCloseParen($saResult, $iParenScan);
		if ($iCloseParen === -1) {
      $bContinue = false; continue;
    }
    const $saRawParams: string = Str["sub"]($saResult,  $iParenScan + 1,  $iCloseParen - $iParenScan - 1);
		const $saTrimmedParams: string = $saRawParams.trim();
		let $saTypedParams: string = "";

		if (Str["len"]($saTrimmedParams) > 0) {
      const $aParts: any[] = Str["split"]($saTrimmedParams,  ",");
			const $iPartsCount: number = $aParts.length;
			let $aTypedParts: any[] = [];
			for (let $iP = 0; $iP < $iPartsCount; $iP = $iP + 1) {
        const $saRawPart: string = $aParts[$iP].trim();
				let $saTypedPart: string = $saRawPart;
				if (Str["len"]($saRawPart) > 0 && Str["indexOf"]($saRawPart,  ":") === -1) {
          $saTypedPart = $saRawPart + ": any";
        }
        $aTypedParts.push( $saTypedPart);
      }
      $saTypedParams = $aTypedParts.join( ", ");
    }
    const $saBefore: string = Str["sub"]($saResult,  0,  $iParenScan + 1);
		const $saAfter: string = Str["sub"]($saResult,  $iCloseParen,  Str["len"]($saResult) - $iCloseParen);

		$saResult = $saBefore + "" + $saTypedParams + "" + $saAfter;
		$iOffset = $iParenScan + 1 + Str["len"]($saTypedParams) + 1;
  }
  return $saResult;
};
